Add Ethereum verifiable signer

rjwebb opened the pull request:
Closes #223, #224.

This implements EIP-712 signing and verification by adding a new Codec `eip712`, which encodes the session and action with EIP-712, then ABI encodes `Message<Session|Action>`. I don't think this is totally right from an architectural POV (see questions) but it does sign and verify correctly.

Questions:
- The `encode` function takes the message as input and nothing else, which means currently it doesn't have any way of passing in the EIP-712 domain. What would be a clean way to do this? I'm guessing @raykyri and @joeltg have ideas... 
- Some of the fields in the action eip-712 type (duration, session) are not in the Action object, again, maybe we need to have some sort of "context" object passed in (but this would break the very clean and coherence signed-cid design) or the encoding actually needs to happen somewhere else, or maybe we need to add some new fields to Message/Action

To do:
- [x] We can create an EIP-712 encoded session with `getSession` and verify it in a smart contract
- [x] We can generate a signed CID and verify it in a smart contract
- [x] We can verify that an Action or Session's content matches a CID in a smart contract - this should be easy, it would involve constructing the CID object inside solidity but we can hardcode the particular codec and digest used.

Changes:
- Added a new EIP712Signer in the `chain-ethereum-eip712` module. This has been added to the general signers tests in `packages/interfaces/signers.test.ts`.
- Added a test to the signers test to sign and verify an action.
- Added a smart contract with tests in the `chain-ethereum-eip712-contract` module. We are using hardhat as a test environment. These tests have been added to CI. Hardhat only supports CommonJS so these tests are in a separate module to the rest of the EIP712Signer.
- Added an `eip712` codec to Canvas which encodes data using EIP712 + ABI
- Added a `none` digest to Canvas which skips the digest stage

### Interface/protocol changes:

Right now we return the secp256k1 compressed public key in the uri (this is the default behaviour of `secp256k1.getPublicKey(privateKey)`). This works fine if we want to check signatures using `secp256k1.verify(message, signature publicKey)` but doesn't work if we are using a verification method that works by recovering the publicKey such as `ecrecover`. We should modify `Secp256k1Signer` so that it returns the **uncompressed** public key in the URI. This should work seamlessly with existing code that uses `secp256k1.verify` (this is confirmed in the tests).

The signature returned by `Secp256k1Signer` does not include a recovery bit, this is fine if you want to just call `secp256k1.verify` because it can try both values - if we want to run something like `ecrecover` we would have to call it twice with each value of the recovery bit. 

### BREAKING CHANGES

~~This PR modifies the call to `secp256k1.sign` in Secp256k1Signer and the call to `secp256k1.verify` to add the `{ prehash: true }` option. This is the same as calling `sha256(...)` on the signed value before it is signed/verified (we could also do this if you think it's more obvious). Currently we are actually not signing hashed messages, so the payload is *truncated*. This means that the signature for CIDs that have the same first 32 bytes are the same.~~ I've made a separate issue to discuss this: https://github.com/canvasxyz/canvas/pull/226 This change has been moved out of this PR into a separate PR: https://github.com/canvasxyz/canvas/pull/235

https://github.com/canvasxyz/canvas/pull/226/files#diff-bff0de8ddb0b9256ac0ac329c1da46ea45b78a4c758797e19ab9858553784859R36

https://github.com/canvasxyz/canvas/pull/226/files#diff-e01002bcc036bca01cb20e116cb1c3b9071d40bf8d78bb59621ec1a059581f61R55

raykyri commented on packages/chain-ethereum-verifiable/src/EIP712VerifiableSigner.ts:
```diff
@@ -0,0 +1,170 @@
+} from "./types.js"
+import { assert, signalInvalidType, validateSessionData, parseAddress, addressPattern, generateSalt } from "./utils.js"
+
+export interface EIP712VerifiableSignerInit {
```
i think we can just call this EIP712Signer across the board now

raykyri commented on packages/chain-ethereum-verifiable/src/EIP712VerifiableSigner.ts:
```diff
@@ -0,0 +1,170 @@
+		this.#ethersSigner = init.signer ?? Wallet.createRandom()
+		this.sessionDuration = init.sessionDuration ?? null
+		this.chainId = init.chainId ?? 1
+		this.verifyingContract = init.verifyingContract ?? "0x1c56346cd2a2bf3202f771f50d3d14a367b48070"
```
what's this contract?

raykyri commented on packages/chain-ethereum-verifiable/src/EIP712VerifiableSigner.ts:
```diff
@@ -0,0 +1,170 @@
+	public constructor(init: EIP712VerifiableSignerInit = {}) {
+		this.#ethersSigner = init.signer ?? Wallet.createRandom()
+		this.sessionDuration = init.sessionDuration ?? null
+		this.chainId = init.chainId ?? 1
```
for chainId and verifyContract, shouldn't the default be to leave them out? they are optional in the spec https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator

rjwebb replied:
yes that's right

rjwebb replied:
ok so re chainId, if it is not provided then how do we format the ethereum address? would it be something like `eip155::0x212421214...` or `eip155:null:0x212421214...` etc?

raykyri commented on packages/chain-ethereum-verifiable/src/types.ts:
```diff
@@ -0,0 +1,30 @@
+
+export type EIP712VerifiableSessionData = {
+	signature: Uint8Array
+	domain: EIP712Domain
```
i think domain (as in document.location, not EIP712 Domain) can just be left out of the signer, it's only in the SIWE signer because the SIWE standard asks for a domain

in metamask, if you ask for a SIWE signature with an invalid/incorrect domain, the SIWE "log in to X website" interface doesn't show up, and it just shows up as text to be signed

rjwebb replied:
oh my bad! i mistook the "domain" field in the SIWESigner for the EIP712 domain. yeah this makes sense, i'll removei t

rjwebb replied:
ah so this is kind of confusing, but the reason why i included the EIP712Domain in the session data object was so that inside verifySession we could call verifyTypedData using the same domain values that were used to sign the session

raykyri replied:
would it make sense to leave it out of the EIP712 signer altogether? ideally we would not even ask the dev for a domain when they configure the signer

raykyri replied:
closing this, i think it'll come up again during review/acceptance testing if it's an issue

raykyri commented on packages/chain-ethereum-verifiable/src/types.ts:
```diff
@@ -0,0 +1,30 @@
+	duration: number | null
+}
+
+export const eip712TypeDefinitionsForSession = {
```
maybe just `eip712TypeDefinitions`?

raykyri commented on packages/signed-cid/src/eip712Codec.ts:
```diff
@@ -0,0 +1,82 @@
+		types.push("string")
+		values.push(key)
+		if (typeof args[key] === "string") {
+			if (args[key].match(/^0x[0-9a-fA-F]*$/)) {
```
this should match exactly 40 chars for addresses - it's possible someone will create an action with a txid or other 0x hexstring.

i wonder if for consistency we should just encode any 0x... hexstring as `bytes`, and specifically those of length 40 as `address`, which eliminates one source of ambiguity. it's still a potential footgun but much less of a surprising one in that case

raykyri commented on packages/signed-cid/src/eip712Codec.ts:
```diff
@@ -0,0 +1,82 @@
+				{ name: "args", type: "bytes" },
+				{ name: "address", type: "address" }, // the address that is delegated-signing the action
+				// TODO: how is this passed into encode?
+				// { name: "session", type: "address" }, // the burner address that is directly signing the action. this could be omitted in the future, but let's include and then revisit during code review
```
we can remove all these comments lol

raykyri commented on packages/signed-cid/src/eip712Codec.ts:
```diff
@@ -0,0 +1,82 @@
+				{ name: "blockhash", type: "string" }, // may be "" if no blockhash
+				{ name: "timestamp", type: "uint256" }, // this is actually overkill at uint24 is enough, but we can revisit during code review
+				// TODO: how is this passed into encode?
+				// { name: "duration", type: "uint256" },
```
no `duration` for action, maybe i copied it into the session type definitions by accident? i fixed this in the latest update to hicommonwealth/commonwealth#6064

raykyri commented on packages/ethereum-contracts/contracts/ECDSA_Verify.sol:
```diff
@@ -0,0 +1,112 @@
+//SPDX-License-Identifier: Unlicense
```
where's this adapted from - did you write part/all of the comment below?

if we're just using the adapted version to avoid installing the OZ package, we might be better off just importing the package

rjwebb replied:
Yes I wrote the comment below.

OK I figured out a much cleaner implementation that imports more of the OZ code instead of just inlining it.

raykyri replied:
Great!

raykyri commented on packages/ethereum-contracts/contracts/CID.sol:
```diff
@@ -0,0 +1,55 @@
+//SPDX-License-Identifier: Unlicense
```
where's this and the Varint library from? i can see references to protobuf/multiformats encoders around github, and this seems to reuse parts of those.

rjwebb replied:
I've deleted the `Varint.sol` library because we can just import the library from here https://github.com/celestiaorg/protobuf3-solidity-lib (I have updated the CID implementation to just import from this library).

The CID code was written by me, I wrote it by translating the CID implementation in js-multiformats (https://github.com/multiformats/js-multiformats/blob/master/src/cid.ts#L447) into solidity.

raykyri commented on packages/ethereum-contracts/contracts/EIP712_Canvas.sol:
```diff
@@ -0,0 +1,178 @@
+
+contract EIP712_Canvas {
+
+    struct Session {
+        address address_;
+        string blockhash_;
+        uint256 duration;
+        string publicKey;
+        uint256 timestamp;
+    }
+
+    struct Action {
+        address address_;
+        bytes args;
+        string blockhash_;
+        string name;
+        uint256 timestamp;
+    }
```
i suppose it's hard to avoid `address_` and `blockhash_` as identifiers?

rjwebb replied:
Yeah, `address` is a keyword which can't be used elsewhere. `blockhash` is the name of a built in function but we can actually use it as a variable, so I've replaced that.

raykyri commented:
I think we can also merge `@canvas-js/chain-ethereum-eip712` into `@canvas-js/chain-ethereum` - to just have SIWESigner.ts and EIP712Signer.ts sitting next to each other as separate imports

rjwebb commented:
@raykyri The comment got buried because I deleted the `Varint.sol` file, but the varint code comes from here https://github.com/celestiaorg/protobuf3-solidity-lib (I have updated the CID implementation to just import from this library).

The CID code was written by me, I wrote it by translating the CID implementation in js-multiformats (https://github.com/multiformats/js-multiformats/blob/master/src/cid.ts#L447) into solidity

raykyri commented:
Oh you made the prehash change here!

We can keep it in, and release 0.9.0 (1.0rc?) when it’s merged.

raykyri commented:
- Can we use the ethers Abicoder instead of web3.eth.abi.encodeParameters? https://docs.ethers.org/v5/api/utils/abi/coder/
- We should add the app's topic to the domain separator, and ensure that tests pass with the topic passed in + also that tests fail when trying to verify a signature for the wrong topic.
- We should add a `Test.sol` contract to the library that performs an action offchain and then submits it onchain. This could just be renamed CID_External since that's just used for testing right now. For example we could add a method like this:

```
string topic = "xyz.canvas.myapp"

function claimUpvoted(uint256 clock, string[] parents, address sender, uint256 postId, ...)  {
    string blockhash = "";
    uint256 duration = 14 * DAYS;
    // ...

    // ... then construct the session, session message, and action message:
    Session memory session = Session(addrs, blockhash, duration, publicKey, timestamp);
    SessionMessage memory sessionMessage = SessionMessage(clock, parents, payload, topic);
    ActionMessage memory actionMessage = ...

    if (!verifySession(session)) { return false; }
    if (!verifySessionMessage(sessionMessage)) { return false; }
    if (!verifyActionMessage(actionMessage)) { return false; }

    // Now, increase a counter stored on this contract by +1, and
    // save the hash of the action in a mapping on the contract's storage,
    // so someone can't submit the same action twice.
}
```

The tests should verify that there is a valid provenance of interactions leading up to the action, i.e. verify both the Session and the ActionMessage. Verifying the SessionMessage seems trivial to add once both those are done, but it would be interesting to see how much gas cost it adds.

raykyri reviewed (commented):
Second round of comments, I haven't yet looked at did:key/did:eth(r)/eth: but this covers everything else.

rjwebb commented on packages/ethereum-contracts/contracts/CID_External.sol:
```diff
@@ -0,0 +1,13 @@
+
+import "./CID.sol";
+
+contract CID_External {
```
To be honest I would rather external devs ignore this file, the only reason why it needs to exist is that hardhat doesn't have a way to directly import a library and call its functions, the library has to first be imported by a contract. I think at the very least `CID_Example` is a less confusing name than `CID_External` so I am happy to rename it to that. Also it might be a good idea to put it in a separate directory like `.../contracts/test`.

rjwebb replied:
What about `CID_Fixture.sol`?

rjwebb commented on packages/ethereum-contracts/contracts/Contract_Example.sol:
```diff
@@ -0,0 +1,64 @@
+
+  function claimUpvoted(
+    address expectedAddress,
+    bytes memory sessionSignature,
```
Yes if the `Session` object has an `authorizationData` field

rjwebb replied:
So we could also write this to expect the `Session` struct to have an `authorizationData` field, then use the signature inside authorizationData instead of passing the signature in separately. This is correct I think, it would just be mirroring what `verifySession` does.

raykyri commented on packages/ethereum-contracts/contracts/Contract_Example.sol:
```diff
@@ -0,0 +1,63 @@
+      "Action must have been signed by a session that has not expired"
+    );
+
+    // validate the action name
```
Also some todos remaining here

rjwebb commented on packages/signed-cid/src/Secp256k1Signer.ts:
```diff
@@ -31,7 +33,7 @@ export class Secp256k1Signer<T = any> implements Signer<T> {
 	public sign(value: T, options: { codec?: string | Codec; digest?: string | Digest } = {}): Signature {
 		const cid = getCID(value, options)
-		const signature = secp256k1.sign(cid.bytes, this.#privateKey).toCompactRawBytes()
+		const signature = secp256k1.sign(cid.bytes, this.#privateKey, { prehash: true }).toCompactRawBytes()
```
PR is here for the record https://github.com/canvasxyz/canvas/pull/235

raykyri commented on packages/ethereum-contracts/contracts/EIP712_Canvas.sol:
```diff
@@ -0,0 +1,274 @@
+        bytes32 digest = _hashTypedDataV4(hashSession(session), name);
+
+        return ECDSA.recover(digest, session.authorizationData.signature) == expectedAddress;
+    }
```
verifySessionWithPubkey

rjwebb replied:
tempted to not implement this until we actually have some code that needs it

raykyri commented:
Looks good!

There are some other tests we should add, and maybe we'll want to support different DIDs or CAIP formatting for the Ethereum address, but those can be handled separately.
